presto-orc/src/test/java/com/facebook/presto/orc/TestOrcWriter.java

```diff
@@ -0,0 +1,118 @@
+import static java.lang.Math.toIntExact;
+
+public class TestOrcWriter
+{
+    @Test
+    public void testWriteOutputStreamsInOrder()
+            throws IOException
+    {
+        TempFile tempFile = new TempFile();
+        OrcWriter writer = createOrcWriter(
+                new OutputStreamSliceOutput(new FileOutputStream(tempFile.getFile())),
+                    ImmutableList.of("test1", "test2", "test3", "test4", "test5"),
+                    ImmutableList.of(VARCHAR, VARCHAR, VARCHAR, VARCHAR, VARCHAR),
+                    NONE,
+                    new DataSize(32, MEGABYTE),
+                    ORC_STRIPE_SIZE,
+                    ORC_STRIPE_SIZE,
+                    ORC_ROW_GROUP_SIZE,
+                    new DataSize(32, MEGABYTE),
+                    ImmutableMap.of(),
+                    HIVE_STORAGE_TIME_ZONE,
+                    true);
+
+        // write down some data with unsorted streams
+        String[] data = new String[]{"a", "bbbbb", "ccc", "dd", "eeee"};
+        Block[] blocks = new Block[data.length];
+        int entries = 65536;
+        BlockBuilder blockBuilder = VARCHAR.createBlockBuilder(new BlockBuilderStatus(), entries);
+        for (int i = 0; i < data.length; i++) {
+            byte[] bytes = data[i].getBytes();
+            for (int j = 0; j < entries; j++) {
+                // force to write different data
+                bytes[0] = (byte) ((bytes[0] + 1) % 128);
+                blockBuilder.writeBytes(Slices.wrappedBuffer(bytes, 0, bytes.length), 0, bytes.length);
+                blockBuilder.closeEntry();
+            }
+            blocks[i] = blockBuilder.build();
+            blockBuilder = blockBuilder.newBlockBuilderLike(new BlockBuilderStatus());
+        }
+
+        writer.write(new Page(blocks));
+        writer.close();
+
+        // read the footer and verify the streams are ordered by size
+        DataSize dataSize = new DataSize(1, MEGABYTE);
+        MetadataReader metadataReader = new OrcMetadataReader();
+        OrcDataSource orcDataSource = new FileOrcDataSource(tempFile.getFile(), dataSize, dataSize, dataSize, true);
+        Footer footer = new OrcReader(orcDataSource, metadataReader, dataSize, dataSize, dataSize).getFooter();
+
+        for (StripeInformation stripe : footer.getStripes()) {
+            // read the footer
+            byte[] tailBuffer = new byte[toIntExact(stripe.getFooterLength())];
+            orcDataSource.readFully(stripe.getOffset() + stripe.getIndexLength() + stripe.getDataLength(), tailBuffer);
+            try (InputStream inputStream = new OrcInputStream(orcDataSource.getId(), Slices.wrappedBuffer(tailBuffer).getInput(), Optional.empty(), new AggregatedMemoryContext())) {
+                StripeFooter stripeFooter = metadataReader.readStripeFooter(footer.getTypes(), inputStream);
+
+                // ignore 5 index streams
+                assertGreaterThan(stripeFooter.getStreams().size(), 5);
```

Comment:
Use `StripeReader.isIndex()` instead (you will need to make it package protected)